src/network/serialize.rs

```diff
@@ -19,13 +19,100 @@
 //! It also defines (de)serialization routines for many primitives.
 //!
 
+use std::error;
+use std::fmt;
+use std::io;
 use std::io::{Cursor, Read, Write};
 use byteorder::{LittleEndian, WriteBytesExt, ReadBytesExt};
 use hex::encode as hex_encode;
 
+use bitcoin_bech32;
+
 use network::encodable::{ConsensusDecodable, ConsensusEncodable};
+use util::base58;
 use util::hash::Sha256dHash;
-use util;
+
+/// Serialization error
+#[derive(Debug)]
+pub enum Error {
+    /// And I/O error
+    Io(io::Error),
+    /// Base58 encoding error
+    Base58(base58::Error),
+    /// Bech32 encoding error
+    Bech32(bitcoin_bech32::Error),
+    /// Error from the `byteorder` crate
+    ByteOrder(io::Error),
+    /// Network magic was not what we expected
+    BadNetworkMagic(u32, u32),
+    /// Network message was unrecognized
+    BadNetworkMessage(String),
+    /// Parsing error
+    ParseFailed,
+    /// Error propagated from subsystem
+    Detail(String, Box<Error>),
+    /// Unsupported witness version
+    UnsupportedWitnessVersion(u8),
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            Error::Io(ref e) => fmt::Display::fmt(e, f),
+            Error::Detail(ref s, ref e) => write!(f, "{}: {}", s, e),
```

Comment:
I think every variant (once we merge `Detail` into `ParseFailed`) should do something here, and drop the `ref x` match.